docs/ci-cd-environment/job-isolation.md

```diff
@@ -0,0 +1,52 @@
+---
+Description: This guide describes how jobs can be isolated in a self-hosted environment.
+---
+
+# Self-hosted job environment isolation
+!!! beta "Self-hosted agents - closed beta"
+    Self-hosted agents are in closed beta. If you would like to run Semaphore agents on your infrastructure, please [contact us and share your use case](https://semaphoreci.com/contact). Our team will get back to you as soon as possible.
+
+Even though some use cases might benefit from jobs sharing the same environment, for a few of them, a pristine environment for each and every job might be needed.
+
+## Using docker containers
+
+Using docker containers is the fastest approach available. Creating, starting, stopping and destroying docker containers is a very fast operation, specially if you cache your docker images in the machine running the agent.
+
+There's two different ways docker containers can be used by the agent:
+
+- the agent itself can run inside a docker container and [disconnect-after-job][disconnect-after-job] can be used to tell it to shutdown after a job is done.
+- jobs can be configured to [use docker images][pipeline yaml], and the self-hosted agent type of your choice. In this scenario, the agent itself doesn't run in a docker container, but the jobs executed by it will. [shutdown-hook-path][shutdown-hook-path] or [disconnect-after-job][disconnect-after-job] are not required if you are doing this, since the jobs are already running in a pristine environment even if the agent is not.
```

Comment:
Active voice:
You can configure jobs to [use docker images][pipeline yaml], and also the self-hosted agent type of your choice.

Active voice and time cohesion:
In this scenario, the agent won't run inside a Docker container. However, the agent will execute the jobs inside the container.


This approach doesn't require [shutdown-hook-path][shutdown-hook-path] or [disconnect-after-job][disconnect-after-job].  

This part, even though I understand what you are going for, looks confusing. I wouldn't add it but if it looks to add it as it is even though it prolongs the sentence:
since the jobs are already running in a pristine environment even if the agent is not.

Reply:
I see the use pristine on several occasions. I do like the word but for this context, it might fit better to use the word `clean`. Pristine is more of a fancy word not used too often while clean is way more common in the industry I believe.